src/Transition_Systems-Weak_Transition_Systems.thy

```diff
@@ -36,6 +36,20 @@ notation weak.constant_bisimilarity (\<open>[\<approx>]\<close>)
 notation weak.simulation_up_to (\<open>weak'_sim\<^bsub>_\<^esub>\<close>)
 notation weak.bisimulation_up_to (\<open>weak'_bisim\<^bsub>_\<^esub>\<close>)
 
+lemma mutual_silent_weak_transitions_up_to_bisimilarity:
+  assumes "p \<Rightarrow>\<lparr>\<tau>\<rparr> p'" and "p' \<approx> q" and "q \<Rightarrow>\<lparr>\<tau>\<rparr> q'" and "q' \<approx> p"
+  shows "p \<approx> q"
+using assms
+proof (coinduction arbitrary: p q p' q' rule: weak.symmetric_up_to_rule [where \<F> = "[\<approx>]"])
+  case (simulation \<alpha> s)
+  from \<open>p \<Rightarrow>\<lparr>\<alpha>\<rparr> s\<close> and \<open>q \<Rightarrow>\<lparr>\<tau>\<rparr> q'\<close> and \<open>q' \<approx> p\<close> obtain s' where "q' \<Rightarrow>\<lparr>\<alpha>\<rparr> s'" and "s' \<approx> s"
+    by (blast elim: weak.bisimilarity.cases)
+  with \<open>q \<Rightarrow>\<lparr>\<tau>\<rparr> q'\<close> have "q \<Rightarrow>\<lparr>\<alpha>\<rparr> s'"
+    by (auto simp add: relcompp_assoc, blast intro: rtranclp_trans)
+  with \<open>s' \<approx> s\<close> show ?case
+    by (fastforce intro: weak.bisimilarity_symmetry_rule)
```

Comment:
With `t ≈ s` (`s' ≈ s`) being replaced by `s ≈ t`, this can be simplified to `by auto`.

Reply:
> With `t ≈ s` (`s' ≈ s`) being replaced by `s ≈ t`, this can be simplified to `by auto`.

Using `try0` I can obtain the following performance assessment:

```
Found proof: by auto (25 ms) 
Found proof: by fastforce (19 ms) 
Found proof: by force (22 ms)
```

The performance of these methods stays more or less the same across several runs of `try0`. Therefore, I think we should stick to `fastforce`. What do you think?

Reply:
I’m not sure. So far, I’ve made my choices of proof methods according to some (not yet publicly documented) simplicity order. This order is partly inspired by the general performance of the different proof methods according to the Isabelle/Isar Reference Manual. According to this order, `auto` is simpler than `fastforce`, which in turn is simpler than `force`. Only where a “simpler” method was obviously slower, I opted for the “not so simple” method.

Of course, it makes sense to take the performance in the concrete case fully into account. That said, we can’t (at least not easily) know in each case which method is faster, as `try0` and, more generally, `try` don’t cover all proof methods. Then again, we might use timing information whenever it is available to us.

As much as I like consistency, I think it’s good to use `fastforce` here and generally switch to taking timing information from tools like `try0` into account. It’s a minor inconsistency, and it has always been a bit up to interpretation what the simplicity order of proof methods should be.

Reply:
Thanks for the elaborated explanation.